src/v/cloud_storage/manifest.h

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "bytes/iobuf.h"
+#include "cloud_storage/types.h"
+#include "cluster/types.h"
+#include "json/json.h"
+#include "model/fundamental.h"
+#include "model/metadata.h"
+#include "s3/client.h"
+#include "seastarx.h"
+#include "tristate.h"
+
+#include <seastar/util/bool_class.hh>
+
+#include <absl/container/btree_map.h>
+#include <rapidjson/fwd.h>
+
+#include <compare>
+#include <iterator>
+
+namespace cloud_storage {
+
+struct serialized_json_stream {
+    ss::input_stream<char> stream;
+    size_t size_bytes;
+};
+
+enum class manifest_type {
+    topic,
+    partition,
+};
+
+class base_manifest {
+public:
+    /// Update manifest file from input_stream (remote set)
+    virtual ss::future<> update(ss::input_stream<char>&& is) = 0;
+
+    /// Serialize manifest object
+    ///
+    /// \return asynchronous input_stream with the serialized json
+    virtual serialized_json_stream serialize() const = 0;
+
+    /// Manifest object name in S3
+    virtual remote_manifest_path get_manifest_path() const = 0;
+
+    /// Get manifest type
+    virtual manifest_type get_manifest_type() const = 0;
+
+    /// Compare two manifests for equality
+    bool operator==(const base_manifest& other) const = default;
+};
+
+/// Manifest file stored in S3
+class manifest final : public base_manifest {
+public:
+    struct segment_meta {
+        bool is_compacted;
+        size_t size_bytes;
+        model::offset base_offset;
+        model::offset committed_offset;
+
+        auto operator<=>(const segment_meta&) const = default;
+    };
+    using key = segment_name;
+    using value = segment_meta;
+    using segment_map = absl::btree_map<key, value>;
+    using const_iterator = segment_map::const_iterator;
+
+    /// Create empty manifest that supposed to be updated later
+    manifest();
+
+    /// Create manifest for specific ntp
+    explicit manifest(model::ntp ntp, model::revision_id rev);
+
+    /// Manifest object name in S3
+    remote_manifest_path get_manifest_path() const override;
+
+    /// Segment file name in S3
+    remote_segment_path get_remote_segment_path(const segment_name& name) const;
+
+    /// Get NTP
+    const model::ntp& get_ntp() const;
+
+    // Get last offset
+    const model::offset get_last_offset() const;
```

Comment:
why return a `const value` doesn't make sense. just return the `model::offset` it's an int64